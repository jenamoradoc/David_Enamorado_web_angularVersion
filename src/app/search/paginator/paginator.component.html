<div class="paginator">
  <div class="per-page">
    <label for="per-page">Items per page: </label>
    <select name="per-page" id="per-page" [formControl]="changeSizePageControl" #perPage>
      <option *ngFor="let pageSize of pageSizeOptions" [value]="pageSize">{{pageSize}}</option>
    </select>
  </div>
  <div class="info-page">
    <p>Showing {{ from }}-{{ to }} of {{ totalCount }} results</p>
  </div>
  <div class="paginator-actions">
    <button class="btn btn-paginator" (click)="onPrev()" [disabled]="to === totalCount"> < </button>
    <button class="btn btn-paginator" (click)="onNext()" [disabled]="to === totalCount "> > </button>
  </div>
</div>
