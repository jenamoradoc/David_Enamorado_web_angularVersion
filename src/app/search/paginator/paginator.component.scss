.paginator {
    margin-top: 10px;
    display: flex;
    align-items: center;
    width: 100%;

    .per-page {
        margin-right: 1rem;
        font-size: 0.825rem;
        color: #999;
    }

    .info-page {
        flex: 1;
        text-align: center;
        font-size: 0.825rem;
        color: #999;
    }

    .btn-paginator {
        margin: 0 5px;
        padding: 5px 10px;
        border: 1px solid #ccc;
        border-radius: 5px;
        transition: all 0.3s ease-in-out;
        font-weight: 900;
        color: #999;

        &:hover {
            background-color: #ccc;
        }
    }
}