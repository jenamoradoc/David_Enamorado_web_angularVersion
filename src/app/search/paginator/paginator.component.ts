import { outputAst } from '@angular/compiler';
import { Component, EventEmitter, Input, OnInit, Output } from '@angular/core';
import { FormControl } from '@angular/forms';
import { distinctUntilChanged, startWith, tap } from 'rxjs/operators';

@Component({
  selector: 'app-paginator',
  templateUrl: './paginator.component.html',
  styleUrls: ['./paginator.component.scss'],
})
export class PaginatorComponent {
  changeSizePageControl = new FormControl<number>(10, { nonNullable: true });

  @Input() pageSizeOptions = [10, 20, 50, 100];
  @Input() totalCount = 0;
  @Output() onChangePage = new EventEmitter<number>();
  @Output() onChangePageSize = this.changeSizePageControl.valueChanges.pipe(
    distinctUntilChanged(),
  );
  // reset paginator
  private _page = 1;
  @Input() set page(value: number) {
    this._page = value;
  }

  get page() {
    return this._page;
  }

  get from() {
    return (this._page - 1) * this.changeSizePageControl.value + 1;
  }

  get to() {
    return Math.min(
      this._page * this.changeSizePageControl.value,
      this.totalCount
    );
  }

  onPrev() {
    if (this._page > 1) {
      this._page -= 1;
      this.onChangePage.emit(this._page);
    }
  }

  onNext() {
    if (this._page < this.changeSizePageControl.value) {
      this._page += 1;
      this.onChangePage.emit(this._page);
    }
  }
}
