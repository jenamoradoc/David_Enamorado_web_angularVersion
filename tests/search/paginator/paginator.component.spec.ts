import { ComponentFixture, TestBed } from '@angular/core/testing';
import { ReactiveFormsModule } from '@angular/forms';

import { PaginatorComponent } from '../../../src/app/search/paginator/paginator.component';

describe('PaginatorComponent', () => {
  let component: PaginatorComponent;
  let fixture: ComponentFixture<PaginatorComponent>;
  let compiled: HTMLElement;

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      declarations: [PaginatorComponent],
      imports: [ReactiveFormsModule],
    }).compileComponents();

    fixture = TestBed.createComponent(PaginatorComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
    compiled = fixture.nativeElement;
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should match snapshot', () => {
    expect(compiled).toMatchSnapshot();
  });

  it('should change page size', () => {
    const selects = compiled.querySelectorAll('select');
    const select = selects[0];
    select.value = '20';
    select.dispatchEvent(new Event('change'));
    const pageSize = +component.changeSizePageControl.value;
    expect(pageSize).toBe(20);
  });

  it('should backward/forward page', () => {
    const buttons = compiled.querySelectorAll('button');
    const prevBtn = buttons[0];
    const nextBtn = buttons[1];
    nextBtn.click();
    nextBtn.dispatchEvent(new Event('click'));
    expect(component.page).toBe(2);

    prevBtn.click();
    prevBtn.dispatchEvent(new Event('click'));
    expect(component.page).toBe(1);
  });
  
  it('should emit onChangePage', () => {
    jest.spyOn(component.onChangePage, 'emit');
    component.onNext();
    expect(component.onChangePage.emit).toHaveBeenCalledWith(2);

    component.onPrev();
    expect(component.onChangePage.emit).toHaveBeenCalledWith(1);
  });
});
