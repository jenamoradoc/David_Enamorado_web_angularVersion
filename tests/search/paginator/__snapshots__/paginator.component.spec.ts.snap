// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PaginatorComponent should match snapshot 1`] = `
<div
  id="root1"
>
  <div
    class="paginator"
  >
    <div
      class="per-page"
    >
      <label
        for="per-page"
      >
        Items per page: 
      </label>
      <select
        class="ng-untouched ng-pristine ng-valid"
        id="per-page"
        name="per-page"
      >
        <option
          value="10"
        >
          10
        </option>
        <option
          value="20"
        >
          20
        </option>
        <option
          value="50"
        >
          50
        </option>
        <option
          value="100"
        >
          100
        </option>
        
      </select>
    </div>
    <div
      class="info-page"
    >
      <p>
        Showing 1-0 of 0 results
      </p>
    </div>
    <div
      class="paginator-actions"
    >
      <button
        class="btn btn-paginator"
        disabled=""
      >
         &lt; 
      </button>
      <button
        class="btn btn-paginator"
        disabled=""
      >
         &gt; 
      </button>
    </div>
  </div>
</div>
`;
